### PR TITLE
PRO-1289: Exchange Token Rate

### PR DESCRIPTION
**Pull request by Jineshdarjee**
## Description

1. Load the token list for all the networks.
2. Read the USD rates information for the whole token list from the reference service i.e. CoinGecho.
3. Read the USD rates information for the whole token list from the Etherspot Rates Service.
4. Compare each rate for each token from the CoinGecho Service and Etherspot Rates Service.
5. The rates from Etherspot and the CoinGecho services deviate by 5%, then the test case should be considered a fail.

### CODE REVIEW

File: test/specs/mainnet/exchangeRate/Smoke_ExchangeRate.spec.js [hunk 1]
```diff
@@ -0,0 +1,188 @@
...
+  "fantom",
+  "aurora",
+  "avalanche",
+  "arbitrum-nova",
+  "moonbeam",
+  "celo",
+  "fuse",
+];
+
+describe("Compare the Token Rates of the Etherspot and CoinGecho Services.", () => {
```
Comment (IAmKio) [inline test/specs/mainnet/exchangeRate/Smoke_ExchangeRate.spec.js:41]:
```suggestion
describe("Compare the Token Rates of the Etherspot and CoinGecko Services", () => {
```

File: test/specs/mainnet/exchangeRate/Smoke_ExchangeRate.spec.js [hunk 2]
```diff
@@ -0,0 +1,188 @@
...
+  for (let n = 0; n < network_etherspot.length; n++) {
+    it("Validate the Token Rates of the Etherspot and CoinGecho Services.", async () => {
```
Comment (IAmKio) [inline test/specs/mainnet/exchangeRate/Smoke_ExchangeRate.spec.js:43]:
```suggestion
    it("Validate the Token Rates of the Etherspot and CoinGecko Services", async () => {
```

File: test/specs/mainnet/exchangeRate/Smoke_ExchangeRate.spec.js [hunk 3]
```diff
@@ -0,0 +1,188 @@
...
+        );
+      } catch (e) {
+        assert.fail("The smart wallet address is not calculated successfully.");
+      }
+
+      // Get the token addresses and it's rate from the CoinGecho
+      try {
+        responsesCoidList = await axios.get(
+          "https://api.coingecko.com/api/v3/coins/list?include_platform=true"
+        );
```
Comment (IAmKio) [inline test/specs/mainnet/exchangeRate/Smoke_ExchangeRate.spec.js:86]: Remember to `.catch` and fail should the request / response not function correctly

File: test/specs/mainnet/exchangeRate/Smoke_ExchangeRate.spec.js [hunk 4]
```diff
@@ -0,0 +1,188 @@
...
+      try {
+        let rates = await mainNetSdk.fetchExchangeRates(requestPayload);
+        for (let y = 0; y < rates.items.length; y++) {
+          for (let j = 0; j < TOKEN_LIST_ADDRESS_COINGECHO.length; j++) {
+            Helper.wait(4000);
+            let etherspotAddress = rates.items[y].address;
+            console.log("Etherspot Address:", etherspotAddress.toLowerCase());
+            console.log("CoinGecho Address:", TOKEN_LIST_ADDRESS_COINGECHO[j]);
+            if (
+              etherspotAddress.toLowerCase() == TOKEN_LIST_ADDRESS_COINGECHO[j]
```
Comment (IAmKio) [inline test/specs/mainnet/exchangeRate/Smoke_ExchangeRate.spec.js:141]: Remember to use strict equality 

```suggestion
              etherspotAddress.toLowerCase() === TOKEN_LIST_ADDRESS_COINGECHO[j]
```

Comment (IAmKio) [inline test/specs/mainnet/exchangeRate/Smoke_ExchangeRate.spec.js:141]: Are we sure that whatever is in `TOKEN_LIST_ADDRESS_COINGECHO[j]` is all lowercase?

File: test/specs/mainnet/exchangeRate/Smoke_ExchangeRate.spec.js [hunk 5]
```diff
@@ -0,0 +1,188 @@
...
+                  TOKEN_LIST_ID_COINGECHO[j] +
+                  "&per_page=100&page=1&sparkline=true&locale=en"
+              );
+
+              let num1 = responsesCoidMarket.data[0].current_price; // rates of the coingecho
+              let num2 = rates.items[y].usd; // rates of the etherspot
+
+              let deviationPercentage =
+                (Math.abs(num1 - num2) / ((num1 + num2) / 2)) * 100;
+              if (deviationPercentage > 5) {
```
Comment (IAmKio) [inline test/specs/mainnet/exchangeRate/Smoke_ExchangeRate.spec.js:156]: Please make the deviation percentage a top level variable as we may need to change this later.

File: test/specs/mainnet/exchangeRate/Smoke_ExchangeRate.spec.js (hunk 1 above)
Comment (Jineshdarjee) [reply test/specs/mainnet/exchangeRate/Smoke_ExchangeRate.spec.js:41]: Modified.

File: test/specs/mainnet/exchangeRate/Smoke_ExchangeRate.spec.js (hunk 2 above)
Comment (Jineshdarjee) [reply test/specs/mainnet/exchangeRate/Smoke_ExchangeRate.spec.js:43]: Modified.

File: test/specs/mainnet/exchangeRate/Smoke_ExchangeRate.spec.js [hunk 6]
```diff
@@ -0,0 +1,188 @@
...
+          "0x666E17ad27fB620D7519477f3b33d809775d65Fe",
+          "The smart wallet address is not calculated correctly."
+        );
+      } catch (e) {
+        assert.fail("The smart wallet address is not calculated successfully.");
+      }
+
+      // Get the token addresses and it's rate from the CoinGecho
+      try {
+        responsesCoidList = await axios.get(
```
[the remark this answers is not in the document]
Comment (Jineshdarjee) [reply test/specs/mainnet/exchangeRate/Smoke_ExchangeRate.spec.js:84]: Modified

File: test/specs/mainnet/exchangeRate/Smoke_ExchangeRate.spec.js (hunk 3 above)
Comment (Jineshdarjee) [reply test/specs/mainnet/exchangeRate/Smoke_ExchangeRate.spec.js:86]: Modified

File: test/specs/mainnet/exchangeRate/Smoke_ExchangeRate.spec.js [hunk 7]
```diff
@@ -0,0 +1,188 @@
...
+          "https://api.coingecko.com/api/v3/coins/list?include_platform=true"
+        );
+
+        for (let z = 0; z < responsesCoidList.data.length; z++) {
```
[the remark this answers is not in the document]
Comment (Jineshdarjee) [reply test/specs/mainnet/exchangeRate/Smoke_ExchangeRate.spec.js:88]: Modified

File: test/specs/mainnet/exchangeRate/Smoke_ExchangeRate.spec.js [hunk 8]
```diff
@@ -0,0 +1,188 @@
...
+          if (
+            typeof responsesCoidList.data[z].platforms[network_coinGecho[n]] ===
```
[the remark this answers is not in the document]
Comment (Jineshdarjee) [reply test/specs/mainnet/exchangeRate/Smoke_ExchangeRate.spec.js:90]: Modified

File: test/specs/mainnet/exchangeRate/Smoke_ExchangeRate.spec.js [hunk 9]
```diff
@@ -0,0 +1,188 @@
...
+              "string" &&
+            responsesCoidList.data[z].platforms[network_coinGecho[n]] != ""
```
[the remark this answers is not in the document]
Comment (Jineshdarjee) [reply test/specs/mainnet/exchangeRate/Smoke_ExchangeRate.spec.js:92]: Modified

File: test/specs/mainnet/exchangeRate/Smoke_ExchangeRate.spec.js [hunk 10]
```diff
@@ -0,0 +1,188 @@
...
+            TOKEN_LIST_ADDRESS_COINGECHO.push(
+              responsesCoidList.data[z].platforms[network_coinGecho[n]]
+            );
+            TOKEN_LIST_ID_COINGECHO.push(responsesCoidList.data[z].id);
+          }
+        }
+      } catch (e) {
+        assert.fail(
+          "An error is displayed while getting the token addresses and it's rate from the CoinGecho."
+        );
```
[the remark this answers is not in the document]
Comment (Jineshdarjee) [reply test/specs/mainnet/exchangeRate/Smoke_ExchangeRate.spec.js:103]: Modified

File: test/specs/mainnet/exchangeRate/Smoke_ExchangeRate.spec.js [hunk 11]
```diff
@@ -0,0 +1,188 @@
...
+          "An error is displayed while fetching the rate of the token from Etherspot."
+        );
+      }
+
+      // Fetch the token rates of the Etherspot and compare with coinGecho
+      try {
+        let rates = await mainNetSdk.fetchExchangeRates(requestPayload);
+        for (let y = 0; y < rates.items.length; y++) {
+          for (let j = 0; j < TOKEN_LIST_ADDRESS_COINGECHO.length; j++) {
+            Helper.wait(4000);
```
[the remark this answers is not in the document]
Comment (Jineshdarjee) [reply test/specs/mainnet/exchangeRate/Smoke_ExchangeRate.spec.js:136]: Modified

File: test/specs/mainnet/exchangeRate/Smoke_ExchangeRate.spec.js [hunk 12]
```diff
@@ -0,0 +1,188 @@
...
+            let etherspotAddress = rates.items[y].address;
+            console.log("Etherspot Address:", etherspotAddress.toLowerCase());
+            console.log("CoinGecho Address:", TOKEN_LIST_ADDRESS_COINGECHO[j]);
+            if (
+              etherspotAddress.toLowerCase() == TOKEN_LIST_ADDRESS_COINGECHO[j]
+            ) {
+              Helper.wait(4000);
```
[the remark this answers is not in the document]
Comment (Jineshdarjee) [reply test/specs/mainnet/exchangeRate/Smoke_ExchangeRate.spec.js:143]: Modified

File: test/specs/mainnet/exchangeRate/Smoke_ExchangeRate.spec.js (hunk 4 above)
Comment (Jineshdarjee) [reply test/specs/mainnet/exchangeRate/Smoke_ExchangeRate.spec.js:141]: Modified

Comment (Jineshdarjee) [reply test/specs/mainnet/exchangeRate/Smoke_ExchangeRate.spec.js:141]: Actually Yes, The address from the CoinGecho is in lower case only. But still, I have modified the code.

File: test/specs/mainnet/exchangeRate/Smoke_ExchangeRate.spec.js (hunk 5 above)
Comment (Jineshdarjee) [reply test/specs/mainnet/exchangeRate/Smoke_ExchangeRate.spec.js:156]: Modified

File: test/specs/mainnet/exchangeRate/Smoke_ExchangeRate.spec.js [hunk 13]
```diff
@@ -0,0 +1,188 @@
...
+                );
+              }
+              break;
+            }
+          }
+        }
+      } catch (e) {
+        assert.fail(
+          "An error is displayed while comparing the rates of the Etherspot and CoinGecho."
+        );
```
[the remark this answers is not in the document]
Comment (Jineshdarjee) [reply test/specs/mainnet/exchangeRate/Smoke_ExchangeRate.spec.js:184]: Modified

**Comment by Jineshdarjee** [comment]
Once the token addresses of the Etherspot and ConinGecho are compared and below mentioned statement will print in the console else it will fail the test case:
"The rate of the weth token of the Etherspot is 1807.88 and the Coingecho is 1804.84. So rate variation of both tokens is displayed correctly."

File: test/specs/mainnet/exchangeRate/Smoke_ExchangeRate.spec.js [hunk 14]
```diff
@@ -0,0 +1,188 @@
...
+            typeof responsesCoidList.data[z].platforms[network_coinGecho[n]] ===
+              "string" &&
+            responsesCoidList.data[z].platforms[network_coinGecho[n]] != ""
+          ) {
+            TOKEN_LIST_ADDRESS_COINGECHO.push(
+              responsesCoidList.data[z].platforms[network_coinGecho[n]]
+            );
+            TOKEN_LIST_ID_COINGECHO.push(responsesCoidList.data[z].id);
+          }
+        }
```
[the remark this answers is not in the document]
Comment (Jineshdarjee) [reply test/specs/mainnet/exchangeRate/Smoke_ExchangeRate.spec.js:117]: Yes, We can add the .filter. But not getting the data in the new array correctly. So as of now, I have not modified the code. But in the future, I'll update the code with .filter

File: test/specs/mainnet/SendFlow/Smoke_NativeTokenTransaction.spec.js [hunk 15]
```diff
@@ -4,8 +4,13 @@ dotenv.config(); // init dotenv
 import { assert } from "chai";
```
Comment (IAmKio) [inline test/specs/mainnet/SendFlow/Smoke_NativeTokenTransaction.spec.js:4]: Did you upstream something here? This wasn't part of this PR originally

**Review by IAmKio — CHANGES_REQUESTED** [review]
There's still 30+ instances of the mispelling `Coingecko`

File: test/specs/mainnet/SendFlow/Smoke_NativeTokenTransaction.spec.js (hunk 15 above)
Comment (Jineshdarjee) [reply test/specs/mainnet/SendFlow/Smoke_NativeTokenTransaction.spec.js:4]: No, I have not upstreamed here.

**Review by IAmKio — APPROVED** [review]

**Pull request merged**